runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/text/Scanner.kt

```diff
@@ -0,0 +1,153 @@
+     * change and the [handler] is not invoked.
+     *
+     * This method is similar to [requireAndSkip] except that no exception is thrown when none of [literals] are found.
+     *
+     * **Example**:
+     *
+     * ```kotlin
+     * val scanner = Scanner("ianbotsf@somewhere.net")
+     * scanner.optionalAndSkip("@") { username ->
+     *     println(username) // Prints "ianbotsf"
+     * }
+     * // Scanner has now skipped "ianbotsf@"
+     * println(scanner) // Prints "Scanner('somewhere.net')"
+     * ```
+     *
+     * @param literals One or more strings to search for at/after the current position
+     * @param handler The handler to invoke on the substring *up to but not including* the nearest found element of
+     * [literals]. If none of [literals] are found, this handler is not invoked.
+     */
+    public fun optionalAndSkip(vararg literals: String, handler: (String) -> Unit) {
+        findAnyOf(literals)?.let { (literal, startIndex) ->
+            processAndSkip(literal, startIndex, handler)
+        }
+    }
+
+    private fun process(untilIndex: Int, handler: (String) -> Unit) {
+        val captured = text.substring(currentIndex, untilIndex)
+        currentIndex = untilIndex
+        handler(captured)
+    }
+
+    private fun processAndSkip(literal: String, startIndex: Int, handler: (String) -> Unit) {
+        process(startIndex, handler)
+        currentIndex += literal.length
+    }
+
+    /**
+     * If any of [literals] is found at/after the current position, invokes the given [handler] on the substring *up to
+     * but not including* the literal and then advances the current position *past* the literal. If multiple of the
+     * given [literals] are found, the nearest one is processed. If none of [literals] is found, an
+     * [IllegalArgumentException] is thrown.
+     *
+     * This method is similar to [optionalAndSkip] except that an exception is thrown when none of [literals] are found.
+     *
+     * **Example**:
+     *
+     * ```kotlin
+     * val scanner = Scanner("ianbotsf@somewhere.net")
+     * scanner.requireAndSkip("@") { username ->
+     *     println(username) // Prints "ianbotsf"
+     * }
+     * // Scanner has now skipped "ianbotsf@"
+     * println(scanner) // Prints "Scanner('somewhere.net')"
+     * ```
+     *
+     * @param literals One or more strings to search for at/after the current position
+     * @param handler The handler to invoke on the substring *up to but not including* the nearest found element of
+     * [literals].
+     */
+    public fun requireAndSkip(vararg literals: String, handler: (String) -> Unit) {
+        val (literal, startIndex) = requireNotNull(findAnyOf(literals)) { "Cannot find any of ${literals.toList()}" }
+        processAndSkip(literal, startIndex, handler)
+    }
+
+    override fun toString(): String = "Scanner(remainingText='${text.substring(currentIndex)}')"
```

Comment:
would it be useful to also print the text as it was given?

Reply:
I personally don't think so for two reasons:

1. It's likely to be verbose and you could wind up with to-string representations like `"Scanner(text='The quick brown fox jumped over the lazy dogs',remainingText='quick brown fox jumped over the lazy dogs')"`. Seems noisy.
1. The scanner is stateful and forward-only. Any text that was previously processed is effectively out of scope and seems irrelevant in a `toString` output. The only reason this class _doesn't_ discard already-processed text (e.g., by making `text` a `var` and replacing it via `substring`) is for performance.